zephyr/lib/regions_mm.c

```diff
@@ -1,52 +1,742 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
  *
  * Author: Jakub Dabek <jakub.dabek@intel.com>
  */
 
 #include <sof/lib/regions_mm.h>
 
-struct virtual_memory_heap
-	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT];
+/* list of virtual_memory_heap objects created */
+struct list_item vmh_list;
 
 /**
- * @brief Fill vm_heaps array with information from zephyr.
+ * @brief Initialize new heap
  *
- * Virtual memory regions calculated in zephyr are translated here
- * to a struct that will keep all information on current allocations
- * and virtual to physical mappings that are related to heaps.
- * System heap is not a part of this information. It only refers to
- * virtual first heaps.
- * Has to be initialized after calculations for regions is done in zephyr.
+ * Initializes new heap based on region that heap is to be assigned to
+ * and config. The heap size overall must be aligned to physical page
+ * size.
+ * @param cfg Configuration structure with block structure for the heap
+ * @param memory_region_attribute A zephyr defined virtual region identifier
+ * @param core_id Core id of a heap that will be created
+ * @param span Flag deciding if heap can do contiguous allocation.
+ *
+ * @retval ptr to a new heap if successful.
+ * @retval NULL on creation failure.
  */
-static int virtual_heaps_init(void)
+struct virtual_memory_heap *vmh_init_heap(const struct vmh_heap_config *cfg,
+	int memory_region_attribute, int core_id, bool span)
 {
 	struct sys_mm_drv_region *virtual_memory_regions =
 		(struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+	int i;
 
-	for (size_t i = 0;
-		i < CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
-		i++) {
-		vm_heaps[i].virtual_region = &virtual_memory_regions[i];
+	/* Check if we haven't created heap for this region already */
+	if (vmh_get_heap_by_attribute(memory_region_attribute, core_id))
+		return NULL;
 
-		switch (virtual_memory_regions[i].attr)	{
-		case MEM_REG_ATTR_CORE_HEAP:
-			vm_heaps[i].memory_caps =
-				SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP | SOF_MEM_CAPS_CACHE;
-			break;
-		case MEM_REG_ATTR_SHARED_HEAP:
-			vm_heaps[i].memory_caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_HP;
+	struct virtual_memory_heap *new_heap =
+		rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*new_heap));
+
+	if (!new_heap)
+		return NULL;
+
+	list_init(&new_heap->node);
+	struct vmh_heap_config new_config = {0};
+
+	/* Search for matching attribute so we place heap on right
+	 * virtual region.
+	 * In case of core heaps regions we count theme from 0 to max
+	 * available cores
+	 */
+	if (memory_region_attribute == MEM_REG_ATTR_CORE_HEAP) {
+		new_heap->virtual_region = &virtual_memory_regions[core_id];
```

Comment:
Will this API only ever be called from core 0 ?
If yes, then lets check core ID and return an error if core != 0.
If no, then we need to manage ```virtual_memory_regions``` as a shared array object. i.e. is it in uncache region ?

Reply:
I had no intent to limit the core from which this api can be called. During offline talks we mentioned few usecases that could use this api from different core.  The virtual_memory_regions is read only  and calculated in zephyr at buildtime. There should be no changes to virtual_memory_regions at any point. Improvement in zephyr required for that I believe.

Reply:
`virtual_region` should be marked `const` then as suggested by me 3.5 months ago.

Reply:
ok, multicore is good. However array elems that are shared between cores must be cache aligned (i.e. make each elem a multiple of cache line size) or must be using uncache memory. Can you confirm memory usage here.

Reply:
@lyakh Yes and as far as i remember it we need to change in zephyr first. So it is read only - should be const in future. I don't see need for cache alignment if no one should change those "read only" values.

Reply:
ok, read only is fine here for cache so lets be ```const```

Reply:
Generaly i must have confused something here since this return from zephyr is const - changing